kythe/rust/extractor/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![feature(rustc_private)]
+
+extern crate rustc_driver;
+extern crate rustc_errors;
+extern crate rustc_interface;
+extern crate rustc_save_analysis;
+extern crate rustc_session;
+
+use rustc_driver::run_compiler;
+use rustc_driver::{Callbacks, Compilation};
+use rustc_interface::{interface, Queries};
+use rustc_save_analysis::DumpHandler;
+use std::path::PathBuf;
+
+/// Generate a save_analysis in `output_dir`
+///
+/// `rustc_arguments` is a `Vec<String>` containing Rust compiler arguments. The
+/// first element must be an empty string.
+/// The save_analysis JSON output file will be located at
+/// {output_dir}/save-analysis/{crate_name}.json
+pub fn generate_analysis(rustc_arguments: Vec<String>, output_dir: PathBuf) -> Result<(), ()> {
+    let mut callback_shim = CallbackShim::new(output_dir);
+
+    // Enable the output of compilation errors to stderr
+    rustc_driver::install_ice_hook();
+
+    rustc_driver::catch_fatal_errors(|| {
+        run_compiler(&rustc_arguments, &mut callback_shim, None, None)
+    })
+    .map(|_| ())
+    .map_err(|_| ())?;
```

Comment:
This looks like it is swallowing any error in running the compiler.  Is there any chance to log the error at least?  Otherwise it may be quite confusing.

Reply:
https://github.com/kythe/kythe/blob/8127f0e701acbe46ba3fa60691455657486f69b6/kythe/rust/extractor/src/lib.rs#L38 tells the compiler to output any compiler errors to stderr on its own. The only error type that gets returned with an `Err` result is [ErrorReported](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.ErrorReported.html) which is not an actual error type.

Reply:
I think it is still useful to at least log the error, if not propagate it upwards.  The type signature here could change or `ErrorReported` may gain additional useful state.  In general, never replace errors with "".

Reply:
The compiler errors already get logged without needing to manually log messages, and we are not provided with a copy of the compiler errors that get logged. Our function also returns an `Err` result to the caller, letting them know that the process failed. 
https://github.com/rust-lang/rls/blob/dd341d53075411cf0f71c5853c98facda35112ef/rls-rustc/src/lib.rs#L80  
also uses the same method, so there is precedent for this. `ErrorReported` is a bit of a misnomer, because it is not a struct that contains the error that was reported but rather indicates that an error *was* reported during compilation. It is empty struct by design.

Reply:
As a compromise I have made the function return a String error containing information about what error occurred and have made it return `"A compiler error has occurred"` when the compilation fails. This also addresses your comment above about ensuring the first argument is an empty string